Fill in PR Parameters #21940

vedantmgoyal2009 opened the pull request:
- [ ] Have you signed the [Contributor License Agreement](https://cla.opensource.microsoft.com/microsoft/winget-pkgs)?
- [ ] Have you checked that there aren't other open [pull requests](https://github.com/microsoft/winget-pkgs/pulls) for the same manifest update/change?
- [ ] Have you validated your manifest locally with `winget validate --manifest <path>`? 
- [ ] Have you tested your manifest locally with `winget install --manifest <path>`?
- [ ] Does your manifest conform to the [1.0 schema](https://github.com/microsoft/winget-cli/blob/master/doc/ManifestSpecv1.0.md)?

Note: `<path>` is the name of the directory containing the manifest you're submitting.

-----

Fixes #21940

vedantmgoyal2009 commented:
@OfficialEsco Please test this version of YamlCreate.ps1

OfficialEsco commented:
This part does not work at all
```
Does this pull request resolve any issues?
Enter issue number. For example: 21983, 43509
[Y] Yes  [N] No (default is 'N'): 2
Warning: 1 uncommitted change
```

Also you don't actually need the `$PSScriptRoot\..\.github\PULL_REQUEST_TEMPLATE.md` part, 
You can do 
```
if ($Task -eq Yes) {
$PRTemplate += "`n[X] Task"
} else {
$PRTemplate += "`n[ ] Task"
}
```
or maybe you could use the PR Template as a Schema?
```
$PRTemplate = Get-Content -path $PSScriptRoot\..\.github\PULL_REQUEST_TEMPLATE.md
foreach ($Line in $PRTemplate {
  Prompt Line without [ ]
  If yes [X] insert else insert [ ]
} 
```

vedantmgoyal2009 commented:
> This part does not work at all
> 
> ```
> Does this pull request resolve any issues?
> Enter issue number. For example: 21983, 43509
> [Y] Yes  [N] No (default is 'N'): 2
> Warning: 1 uncommitted change
> ```

I have fixed this part.

> Also you don't actually need the `$PSScriptRoot\..\.github\PULL_REQUEST_TEMPLATE.md` part,
> You can do
> 
> ```
> if ($Task -eq Yes) {
> $PRTemplate += "`n[X] Task"
> } else {
> $PRTemplate += "`n[ ] Task"
> }
> ```
> 
> or maybe you could use the PR Template as a Schema?
> 
> ```
> $PRTemplate = Get-Content -path $PSScriptRoot\..\.github\PULL_REQUEST_TEMPLATE.md
> foreach ($Line in $PRTemplate {
>   Prompt Line without [ ]
>   If yes [X] insert else insert [ ]
> } 
> ```

I first tried to take a variable and then use `$PrBody = Get-Content -path $PSScriptRoot\..\.github\PULL_REQUEST_TEMPLATE.md`, then replace it as you mentioned, but GitHub CLI does not take it line by line and what you will get is 
```
- [ ] Have you signed the [Contributor License Agreement](https://cla.opensource.microsoft.com/microsoft/winget-pkgs)?- [ ] Have you checked that there aren't other open [pull requests](https://github.com/microsoft/winget-pkgs/pulls) for the same manifest update/change?- [ ] Have you validated your manifest locally with `winget validate --manifest <path>`? - [ ] Have you tested your manifest locally with `winget install --manifest <path>`?- [ ] Does your manifest conform to the [1.0 schema](https://github.com/microsoft/winget-cli/blob/master/doc/ManifestSpecv1.0.md)?   Note: `<path>` is the name of the directory containing the manifest you're submitting.  -----

```

OfficialEsco commented:
> I have fixed this part.

Yap less confusing now

> I first tried to use $PrBody = Get-Content -path $PSScriptRoot\..\.github\PULL_REQUEST_TEMPLATE.md, then replace it as you mentioned, but GitHub CLI does not take it line by line and what you will get is a string like this, that's why I need to write such a long function.

Took me waaay too long to test gh pr create 🤦‍♂️ You're right, it actually needs a file.. thats dumb

What do you think of something like this? (Testing draft)
```
Function Enter-PR-Parameters {
    Clear-Variable PrBodyContentReply
    $PrBodyContent = Get-Content -Path "$env:USERPROFILE\Github\winget-pkgs\.github\PULL_REQUEST_TEMPLATE.md" 

    ForEach ($_ in ($PrBodyContent | Where-Object { $_ -like '-*[ ]*' })) {
        if ($_ -like "*winget validate*") {
            $PrBodyContentReply += $_.Replace("[ ]","[X]"), "`n"
        } else {
            Write-Host
            Write-Host
            Write-Host -ForegroundColor 'White' $_.TrimStart("- [ ]")
            Write-Host -ForegroundColor 'White' -NoNewline "[Y] Yes  "
            Write-Host -ForegroundColor 'Yellow' -NoNewline "[N] No "
            Write-Host -NoNewline "(default is 'N'): "
            do {
                $keyInfo = [Console]::ReadKey($false)
            } until ($keyInfo.Key)
            if ($keyInfo.Key -eq 'Y') {
                $PrBodyContentReply += $_.Replace("[ ]","[X]"), "`n"
            } else {
                $PrBodyContentReply += $_, "`n"
            }
        }
    }

    Write-Host
    Write-Host
    Write-Host -ForegroundColor 'White' "Does this pull request resolve any issues?"
    Write-Host -ForegroundColor 'White' -NoNewline "[Y] Yes  "
    Write-Host -ForegroundColor 'Yellow' -NoNewline "[N] No "
    Write-Host -NoNewline "(default is 'N'): "
    do {
        $keyInfo = [Console]::ReadKey($false)
    } until ($keyInfo.Key)
    if ($keyInfo.Key -eq 'Y') {
        Write-Host
        Write-Host "Enter issue number. For example`: 21983, 43509"
        $ResolvedIssues = Read-Host -Prompt 'Resolved Issues' | TrimString
        Foreach($i in ($ResolvedIssues.Split(","))) {
            $PrBodyContentReply += "Resolves #$i`n"
        }
    } else {Write-Host}

    Write-Host
    $PrBodyContentReply = ($PrBodyContentReply.Trim() -ne '')
    $PrBodyContentReply

    #Set-Content -Path PrBodyFile -Value $PrBodyContentReply | Out-Null
    #gh pr create --body-file PrBodyFile -f
    #Remove-Item PrBodyFile
}
```

I have this issue:
```
Does this pull request resolve any issues?
[Y] Yes  [N] No (default is 'N'): y
Enter issue number. For example: 21983, 43509
Resolved Issues: 123, 1234

- [X] Have you signed the [Contributor License Agreement](https://cla.opensource.microsoft.com/microsoft/winget-pkgs)?
- [X] Have you checked that there aren't other open [pull requests](https://github.com/microsoft/winget-pkgs/pulls) for the same manifest update/change?
- [X] Have you validated your manifest locally with `winget validate --manifest <path>`?
- [X] Have you tested your manifest locally with `winget install --manifest <path>`?
- [X] Does your manifest conform to the [1.0 schema](https://github.com/microsoft/winget-cli/blob/master/doc/ManifestSpecv1.0.md)?
Resolves #123
Resolves # 1234
```
For no space i have to do `Resolved Issues: 123,1234`

Dunno if its because i've messed with your part

Also i agree the `Test-Manifest` should not be there 😁
Also the script already runs `winget validate`, so i don't think there is a reason for it do do it again so we just check validate automatically .

vedantmgoyal2009 commented:
> Also i agree the `Test-Manifest` should not be there 😁

Fixed in commit 0daf9f0bfda3686ce3a653753d4ac83d46bfdd4c

> For no space i have to do `Resolved Issues: 123,1234`

Fixed in commit 4f24c84993f925dcc73c838f42d734adf989e2a1... Now i think everything is perfect...

> Also the script already runs `winget validate`, so i don't think there is a reason for it do do it again so we just check validate automatically .

You are right... I have removed that line in commit 4dfd2abebfe9d127d6361b195a9d06020c0318c9

```
Function Enter-PR-Parameters {
    Clear-Variable PrBodyContentReply
    $PrBodyContent = Get-Content -Path "$env:USERPROFILE\Github\winget-pkgs\.github\PULL_REQUEST_TEMPLATE.md" 

    ForEach ($_ in ($PrBodyContent | Where-Object { $_ -like '-*[ ]*' })) {
        if ($_ -like "*winget validate*") {
            $PrBodyContentReply += $_.Replace("[ ]","[X]"), "`n"
        } else {
            Write-Host
            Write-Host
            Write-Host -ForegroundColor 'White' $_.TrimStart("- [ ]")
            Write-Host -ForegroundColor 'White' -NoNewline "[Y] Yes  "
            Write-Host -ForegroundColor 'Yellow' -NoNewline "[N] No "
            Write-Host -NoNewline "(default is 'N'): "
            do {
                $keyInfo = [Console]::ReadKey($false)
            } until ($keyInfo.Key)
            if ($keyInfo.Key -eq 'Y') {
                $PrBodyContentReply += $_.Replace("[ ]","[X]"), "`n"
            } else {
                $PrBodyContentReply += $_, "`n"
            }
        }
    }

    Write-Host
    Write-Host
    Write-Host -ForegroundColor 'White' "Does this pull request resolve any issues?"
    Write-Host -ForegroundColor 'White' -NoNewline "[Y] Yes  "
    Write-Host -ForegroundColor 'Yellow' -NoNewline "[N] No "
    Write-Host -NoNewline "(default is 'N'): "
    do {
        $keyInfo = [Console]::ReadKey($false)
    } until ($keyInfo.Key)
    if ($keyInfo.Key -eq 'Y') {
        Write-Host
        Write-Host "Enter issue number. For example`: 21983, 43509"
        $ResolvedIssues = Read-Host -Prompt 'Resolved Issues' | TrimString
        Foreach($i in ($ResolvedIssues.Split(","))) {
            $PrBodyContentReply += "Resolves #$i`n"
        }
    } else {Write-Host}

    Write-Host
    $PrBodyContentReply = ($PrBodyContentReply.Trim() -ne '')
    $PrBodyContentReply

    #Set-Content -Path PrBodyFile -Value $PrBodyContentReply | Out-Null
    #gh pr create --body-file PrBodyFile -f
    #Remove-Item PrBodyFile
}
```
I am completely in favour of this. It will simplify the code much but for any first-time contributor who will be using this script, won't he will be going through a kind of weird prompts?
![image](https://user-images.githubusercontent.com/83997633/128357903-97a2e223-a2ae-4294-832f-be81c28ca06a.png)

OfficialEsco commented:
> Fixed in commit 4f24c84... Now i think everything is perfect...

What, I swear I tested that and it had no effect... Probably broke something on my end

> won't he will be going through a kind of weird prompts?

Yeeah it looks ugly, you can make a `elseif` for a custom prompt like the winget validate part, also if user selected yes to test in Sandbox the winget install prompt should be checked automatically with a `elseif`

vedantmgoyal2009 commented:
> > Fixed in commit [4f24c84](https://github.com/Trenly/winget-pkgs/commit/4f24c84993f925dcc73c838f42d734adf989e2a1)... Now i think everything is perfect...
> 
> What, I swear I tested that and it had no effect... Probably broke something on my end

I had tested PR Parameters feature on PS 5.1. What version are you currently using? I don't think it should break. Here is what I have tested. 
![image](https://user-images.githubusercontent.com/83997633/128361647-18cea1da-bdbd-4642-bf50-8c5989a61a9e.png)
![image](https://user-images.githubusercontent.com/83997633/128361578-8e621b03-8dcf-4511-b6fa-6efec02f9037.png)

vedantmgoyal2009 commented:
> Yeeah it looks ugly, you can make a `elseif` for a custom prompt like the winget validate part, also if user selected yes to test in Sandbox the winget install prompt should be checked automatically with a `elseif`

Will be making the changes and committing them soon. Thanks for the suggestion.

OfficialEsco commented:
I use the Default Powershell (5.1.19041.1023) which comes with Windows 10 for compatibility, but our code is a tiny bit different since you haven't implemented the simplified prompts yet 😁 so we'll see where i messed up once you implement it since i believe you need to do the variable part differently👍

vedantmgoyal2009 commented:
> ```
> Function Enter-PR-Parameters {
>     Clear-Variable PrBodyContentReply
>     $PrBodyContent = Get-Content -Path "$env:USERPROFILE\Github\winget-pkgs\.github\PULL_REQUEST_TEMPLATE.md" 
> 
>     ForEach ($_ in ($PrBodyContent | Where-Object { $_ -like '-*[ ]*' })) {
>         if ($_ -like "*winget validate*") {
>             $PrBodyContentReply += $_.Replace("[ ]","[X]"), "`n"
>         } else {
>             Write-Host
>             Write-Host
>             Write-Host -ForegroundColor 'White' $_.TrimStart("- [ ]")
>             Write-Host -ForegroundColor 'White' -NoNewline "[Y] Yes  "
>             Write-Host -ForegroundColor 'Yellow' -NoNewline "[N] No "
>             Write-Host -NoNewline "(default is 'N'): "
>             do {
>                 $keyInfo = [Console]::ReadKey($false)
>             } until ($keyInfo.Key)
>             if ($keyInfo.Key -eq 'Y') {
>                 $PrBodyContentReply += $_.Replace("[ ]","[X]"), "`n"
>             } else {
>                 $PrBodyContentReply += $_, "`n"
>             }
>         }
>     }
> 
>     Write-Host
>     Write-Host
>     Write-Host -ForegroundColor 'White' "Does this pull request resolve any issues?"
>     Write-Host -ForegroundColor 'White' -NoNewline "[Y] Yes  "
>     Write-Host -ForegroundColor 'Yellow' -NoNewline "[N] No "
>     Write-Host -NoNewline "(default is 'N'): "
>     do {
>         $keyInfo = [Console]::ReadKey($false)
>     } until ($keyInfo.Key)
>     if ($keyInfo.Key -eq 'Y') {
>         Write-Host
>         Write-Host "Enter issue number. For example`: 21983, 43509"
>         $ResolvedIssues = Read-Host -Prompt 'Resolved Issues' | TrimString
>         Foreach($i in ($ResolvedIssues.Split(","))) {
>             $PrBodyContentReply += "Resolves #$i`n"
>         }
>     } else {Write-Host}
> 
>     Write-Host
>     $PrBodyContentReply = ($PrBodyContentReply.Trim() -ne '')
>     $PrBodyContentReply
> 
>     #Set-Content -Path PrBodyFile -Value $PrBodyContentReply | Out-Null
>     #gh pr create --body-file PrBodyFile -f
>     #Remove-Item PrBodyFile
> }
> ```

I recently discovered that it will omit the last two lines of the template, aren't they important. If not, I will be cleaning up and testing the Function and commit it.
```
Note: `<path>` is the name of the directory containing the manifest you're submitting.

-----
```

OfficialEsco commented:
Naaaaaah tbh it looks cleaner without the 2 last lines, the note is just to inform how to do `winget install` and `winget validate`
And the `-----` like have no function, and if we type something above it the text will be huge 🤦‍♂️
```
BIG TEXT
-----
```
```
Normal text

-----
```

vedantmgoyal2009 commented:
> Let me know when its implemented :)

Done in commit 9b5cdbded23493fe18b3f513d76371e0ad5db077

OfficialEsco commented on Tools/YamlCreate.ps1:
```diff
@@ -786,6 +786,132 @@ Function Test-Manifest {
+        } elseif ($_ -like "*schema*") {
+            Write-Host
+            Write-Host
+            Write-Host -ForegroundColor 'White' "Does your manifest confirm to the 1.0 schema?"
```
Typo =)
```suggestion
            Write-Host -ForegroundColor 'White' "Does your manifest conform to the 1.0 schema?"
```

OfficialEsco commented:
Uhoh something gone wrong
https://github.com/microsoft/winget-pkgs/pull/23909
You need a function like this somewhere i believe

```
$PrBodyContentReply = ($PrBodyContentReply.Trim() -ne '')
```

vedantmgoyal2009 commented:
Oh! actually, I wrote the code in Untitled.ps1 and copied it here. error in copying the code 🤦‍♂️

vedantmgoyal2009 commented:
https://desktop.githubusercontent.com/releases/2.4.3-539849ed/GitHubDesktopSetup.msi I think it should be removed.

OfficialEsco commented:
> https://desktop.githubusercontent.com/releases/2.4.3-539849ed/GitHubDesktopSetup.msi I think it should be removed.

```
This XML file does not appear to have any style information associated with it. The document tree is shown below.
<Error>
<Code>ResourceNotFound</Code>
<Message>The specified resource does not exist. RequestId:cb182efa-d01e-004f-0a1a-8a5a9d000000 Time:2021-08-05T16:51:13.2113922Z</Message>
</Error>
```

But if you are talking about the GitHub Desktop Machine Wide installer we're waiting for Schema v1.1.0

vedantmgoyal2009 commented:
No.. Download URL is invalid in manifests. So I am removing the manifests.

OfficialEsco commented:
Uhhh i just checked a few URLs and they all fail up to the latest 2.9.0
https://desktop.githubusercontent.com/releases/2.9.0-4806a6dc/GitHubDesktopSetup-x64.msi and https://desktop.githubusercontent.com/releases/2.9.0-4806a6dc/GitHubDesktopSetup-x64.exe

vedantmgoyal2009 commented:
> Uhhh i just checked a few URLs and they all fail up to the latest 2.9.0
> https://desktop.githubusercontent.com/releases/2.9.0-4806a6dc/GitHubDesktopSetup-x64.msi and https://desktop.githubusercontent.com/releases/2.9.0-4806a6dc/GitHubDesktopSetup-x64.exe

What should we do?
Remove all manifests... 
version 2.9.0 is accessible through the browser...
![image](https://user-images.githubusercontent.com/83997633/128391185-a9a57dba-532b-4b52-ad00-518087c32039.png)

OfficialEsco commented:
Yup seems like they changed the URL's and the older versions are still not accessable
https://desktop.githubusercontent.com/github-desktop/releases/2.9.0-4806a6dc/GitHubDesktopSetup-x64.exe
https://desktop.githubusercontent.com/github-desktop/releases/2.9.0-4806a6dc/GitHubDesktopSetup-x64.msi

Guess it dosn't hurt to purge the package

vedantmgoyal2009 commented:
> Yup seems like they changed the URL's and the older versions are still not accessable
> https://desktop.githubusercontent.com/github-desktop/releases/2.9.0-4806a6dc/GitHubDesktopSetup-x64.exe
> https://desktop.githubusercontent.com/github-desktop/releases/2.9.0-4806a6dc/GitHubDesktopSetup-x64.msi
> 
> Guess it dosn't hurt to purge the package

What should I do finally, remove manifests or change URLs? Levvie approved them so fast!

OfficialEsco commented:
Check if inserting /github-desktop/ in the URL works, if not delete them

vedantmgoyal2009 commented:
@Trenly you didn't approve this one?

Trenly commented on Tools/YamlCreate.ps1:
```diff
@@ -786,6 +786,133 @@ Function Test-Manifest {
+Function Enter-PR-Parameters {
+    $PrBodyContent = Get-Content $args[0]
+    ForEach ($_ in ($PrBodyContent | Where-Object { $_ -like '-*[ ]*' })) {
+        if ($_ -like "*CLA*") {
```
Can you make this a [wildcard switch](https://docs.microsoft.com/en-us/powershell/scripting/learn/deep-dives/everything-about-switch?view=powershell-7.1#-wildcard)  instead of if else-if statements?

OfficialEsco replied:
Oooo nowonder my $variable switch didn't work, if only i had google 🤦‍♂️
> I originally tried to do that part with `switch` but I don't understand it >:c

vedantmgoyal2009 replied:
Will try to do it!

OfficialEsco replied:
@vedantmgoyal2009 i've got a working switch now, also fixed the Sandbox question which was not working

vedantmgoyal2009 replied:
Can you give me the code? I will try to fix it.

OfficialEsco commented on Tools/YamlCreate.ps1:
```diff
@@ -786,6 +786,133 @@ Function Test-Manifest {
+                } until ($keyInfo.Key)
+                if ($keyInfo.Key -eq 'Y') {
+                    $PrBodyContentReply += $_.Replace("[ ]","[X]"), "`n"
+                } else {
+                    $PrBodyContentReply += $_, "`n"
+                }
+            }
+        } elseif ($_ -like "*schema*") {
+            Write-Host
+            Write-Host
+            Write-Host -ForegroundColor 'White' "Does your manifest conform to the 1.0 schema?"
+            Write-Host "Reference Link: https://github.com/microsoft/winget-cli/blob/master/doc/ManifestSpecv1.0.md"
+            Write-Host -ForegroundColor 'White' -NoNewline "[Y] Yes  "
+            Write-Host -ForegroundColor 'Yellow' -NoNewline "[N] No "
+            Write-Host -NoNewline "(default is 'N'): "
+            do {
+                $keyInfo = [Console]::ReadKey($false)
+            } until ($keyInfo.Key)
+            if ($keyInfo.Key -eq 'Y') {
+                $PrBodyContentReply += $_.Replace("[ ]","[X]"), "`n"
+            } else {
+                $PrBodyContentReply += $_, "`n"
+            }
+        } else {
+            Write-Host
+            Write-Host
+            Write-Host -ForegroundColor 'White' $_.TrimStart("- [ ]")
+            Write-Host -ForegroundColor 'White' -NoNewline "[Y] Yes  "
+            Write-Host -ForegroundColor 'Yellow' -NoNewline "[N] No "
+            Write-Host -NoNewline "(default is 'N'): "
+            do {
+                $keyInfo = [Console]::ReadKey($false)
+            } until ($keyInfo.Key)
+            if ($keyInfo.Key -eq 'Y') {
+                $PrBodyContentReply += $_.Replace("[ ]","[X]"), "`n"
+            } else {
+                $PrBodyContentReply += $_, "`n"
+            }
+        }
+    }
```
```suggestion
    ForEach ($_ in ($PrBodyContent | Where-Object { $_ -like '-*[ ]*' })) {
            switch -Wildcard ( $_ )
            {
                '*CLA*' {
                    Write-Host
                    Write-Host -ForegroundColor 'White' "Have you signed the Contributor License Agreement (CLA)?"
                    Write-Host "Reference Link: https://cla.opensource.microsoft.com/microsoft/winget-pkgs"
                    Write-Host -ForegroundColor 'White' -NoNewline "[Y] Yes  "
                    Write-Host -ForegroundColor 'Yellow' -NoNewline "[N] No "
                    Write-Host -NoNewline "(default is 'N'): "
                    do {
                        $keyInfo = [Console]::ReadKey($false)
                    } until ($keyInfo.Key)
                    if ($keyInfo.Key -eq 'Y') {
                        $PrBodyContentReply += $_.Replace("[ ]","[X]"), "`n"
                    } else {
                        $PrBodyContentReply += $_, "`n"
                    }
                }
    
                '*open `[pull requests`]*' {
                    Write-Host
                    Write-Host
                    Write-Host -ForegroundColor 'White' "Have you checked that there aren't other open pull requests for the same manifest update/change?"
                    Write-Host "Reference Link: https://github.com/microsoft/winget-pkgs/pulls"
                    Write-Host -ForegroundColor 'White' -NoNewline "[Y] Yes  "
                    Write-Host -ForegroundColor 'Yellow' -NoNewline "[N] No "
                    Write-Host -NoNewline "(default is 'N'): "
                    do {
                        $keyInfo = [Console]::ReadKey($false)
                    } until ($keyInfo.Key)
                    if ($keyInfo.Key -eq 'Y') {
                        $PrBodyContentReply += $_.Replace("[ ]","[X]"), "`n"
                    } else {
                        $PrBodyContentReply += $_, "`n"
                    }
                }
    
                '*winget validate*' {
                    if($?) {
                        $PrBodyContentReply += $_.Replace("[ ]","[X]"), "`n"
                    } else {
                        Write-Host
                        Write-Host
                        Write-Host -ForegroundColor 'Red' "Automatic manifest validation failed. Check your manifest and try again"
                        Write-Host -ForegroundColor 'White' "Have you validated your manifest locally with 'winget validate --manifest <path>'"
                        Write-Host -ForegroundColor 'White' -NoNewline "[Y] Yes  "
                        Write-Host -ForegroundColor 'Yellow' -NoNewline "[N] No "
                        Write-Host -NoNewline "(default is 'N'): "
                        do {
                            $keyInfo = [Console]::ReadKey($false)
                        } until ($keyInfo.Key)
                        if ($keyInfo.Key -eq 'Y') {
                            $PrBodyContentReply += $_.Replace("[ ]","[X]"), "`n"
                        } else {
                            $PrBodyContentReply += $_, "`n"
                        }
                    }
                }
    
                '*tested your manifest*' {
                    if ($SandboxTest -eq '0') {
                        $PrBodyContentReply += $_.Replace("[ ]","[X]"), "`n"
                    } else {
                        Write-Host
                        Write-Host
                        Write-Host -ForegroundColor 'Yellow' "You did not test your Manifest in Windows Sandbox previously."
                        Write-Host -ForegroundColor 'White' "Have you tested your manifest locally with 'winget install --manifest <path>'"
                        Write-Host -ForegroundColor 'White' -NoNewline "[Y] Yes  "
                        Write-Host -ForegroundColor 'Yellow' -NoNewline "[N] No "
                        Write-Host -NoNewline "(default is 'N'): "
                        do {
                            $keyInfo = [Console]::ReadKey($false)
                        } until ($keyInfo.Key)
                        if ($keyInfo.Key -eq 'Y') {
                            $PrBodyContentReply += $_.Replace("[ ]","[X]"), "`n"
                        } else {
                            $PrBodyContentReply += $_, "`n"
                        }
                    }
                }
    
                '*schema*' {
                    Write-Host
                    Write-Host
                    Write-Host -ForegroundColor 'White' "Does your manifest conform to the 1.0 schema?"
                    Write-Host "Reference Link: https://github.com/microsoft/winget-cli/blob/master/doc/ManifestSpecv1.0.md"
                    Write-Host -ForegroundColor 'White' -NoNewline "[Y] Yes  "
                    Write-Host -ForegroundColor 'Yellow' -NoNewline "[N] No "
                    Write-Host -NoNewline "(default is 'N'): "
                    do {
                        $keyInfo = [Console]::ReadKey($false)
                    } until ($keyInfo.Key)
                    if ($keyInfo.Key -eq 'Y') {
                        $PrBodyContentReply += $_.Replace("[ ]","[X]"), "`n"
                    } else {
                        $PrBodyContentReply += $_, "`n"
                    }
                }
    
                Default {
                    Write-Host
                    Write-Host
                    Write-Host -ForegroundColor 'White' $_.TrimStart("- [ ]")
                    Write-Host -ForegroundColor 'White' -NoNewline "[Y] Yes  "
                    Write-Host -ForegroundColor 'Yellow' -NoNewline "[N] No "
                    Write-Host -NoNewline "(default is 'N'): "
                    do {
                        $keyInfo = [Console]::ReadKey($false)
                    } until ($keyInfo.Key)
                    if ($keyInfo.Key -eq 'Y') {
                        $PrBodyContentReply += $_.Replace("[ ]","[X]"), "`n"
                    } else {
                        $PrBodyContentReply += $_, "`n"
                    }
                }
            }
        }
```

OfficialEsco replied:
I think i might have forgotten a `}` above `Write-Host` above `"Does this pull request resolve any issues?"`

OfficialEsco commented on Tools/YamlCreate.ps1:
```diff
@@ -786,6 +786,133 @@ Function Test-Manifest {
+                } until ($keyInfo.Key)
+                if ($keyInfo.Key -eq 'Y') {
+                    $PrBodyContentReply += $_.Replace("[ ]","[X]"), "`n"
+                } else {
+                    $PrBodyContentReply += $_, "`n"
+                }
+            }
+        } elseif ($_ -like "*schema*") {
+            Write-Host
+            Write-Host
+            Write-Host -ForegroundColor 'White' "Does your manifest conform to the 1.0 schema?"
+            Write-Host "Reference Link: https://github.com/microsoft/winget-cli/blob/master/doc/ManifestSpecv1.0.md"
+            Write-Host -ForegroundColor 'White' -NoNewline "[Y] Yes  "
+            Write-Host -ForegroundColor 'Yellow' -NoNewline "[N] No "
+            Write-Host -NoNewline "(default is 'N'): "
+            do {
+                $keyInfo = [Console]::ReadKey($false)
+            } until ($keyInfo.Key)
+            if ($keyInfo.Key -eq 'Y') {
+                $PrBodyContentReply += $_.Replace("[ ]","[X]"), "`n"
+            } else {
+                $PrBodyContentReply += $_, "`n"
+            }
+        } else {
+            Write-Host
+            Write-Host
+            Write-Host -ForegroundColor 'White' $_.TrimStart("- [ ]")
+            Write-Host -ForegroundColor 'White' -NoNewline "[Y] Yes  "
+            Write-Host -ForegroundColor 'Yellow' -NoNewline "[N] No "
+            Write-Host -NoNewline "(default is 'N'): "
+            do {
+                $keyInfo = [Console]::ReadKey($false)
+            } until ($keyInfo.Key)
+            if ($keyInfo.Key -eq 'Y') {
+                $PrBodyContentReply += $_.Replace("[ ]","[X]"), "`n"
+            } else {
+                $PrBodyContentReply += $_, "`n"
+            }
+        }
+    }
```
```suggestion
    ForEach ($_ in ($PrBodyContent | Where-Object { $_ -like '-*[ ]*' })) {
            switch -Wildcard ( $_ )
            {
                '*CLA*' {
                    Write-Host
                    Write-Host -ForegroundColor 'White' "Have you signed the Contributor License Agreement (CLA)?"
                    Write-Host "Reference Link: https://cla.opensource.microsoft.com/microsoft/winget-pkgs"
                    Write-Host -ForegroundColor 'White' -NoNewline "[Y] Yes  "
                    Write-Host -ForegroundColor 'Yellow' -NoNewline "[N] No "
                    Write-Host -NoNewline "(default is 'N'): "
                    do {
                        $keyInfo = [Console]::ReadKey($false)
                    } until ($keyInfo.Key)
                    if ($keyInfo.Key -eq 'Y') {
                        $PrBodyContentReply += $_.Replace("[ ]","[X]"), "`n"
                    } else {
                        $PrBodyContentReply += $_, "`n"
                    }
                }
    
                '*open `[pull requests`]*' {
                    Write-Host
                    Write-Host
                    Write-Host -ForegroundColor 'White' "Have you checked that there aren't other open pull requests for the same manifest update/change?"
                    Write-Host "Reference Link: https://github.com/microsoft/winget-pkgs/pulls"
                    Write-Host -ForegroundColor 'White' -NoNewline "[Y] Yes  "
                    Write-Host -ForegroundColor 'Yellow' -NoNewline "[N] No "
                    Write-Host -NoNewline "(default is 'N'): "
                    do {
                        $keyInfo = [Console]::ReadKey($false)
                    } until ($keyInfo.Key)
                    if ($keyInfo.Key -eq 'Y') {
                        $PrBodyContentReply += $_.Replace("[ ]","[X]"), "`n"
                    } else {
                        $PrBodyContentReply += $_, "`n"
                    }
                }
    
                '*winget validate*' {
                    if($?) {
                        $PrBodyContentReply += $_.Replace("[ ]","[X]"), "`n"
                    } else {
                        Write-Host
                        Write-Host
                        Write-Host -ForegroundColor 'Red' "Automatic manifest validation failed. Check your manifest and try again"
                        Write-Host -ForegroundColor 'White' "Have you validated your manifest locally with 'winget validate --manifest <path>'"
                        Write-Host -ForegroundColor 'White' -NoNewline "[Y] Yes  "
                        Write-Host -ForegroundColor 'Yellow' -NoNewline "[N] No "
                        Write-Host -NoNewline "(default is 'N'): "
                        do {
                            $keyInfo = [Console]::ReadKey($false)
                        } until ($keyInfo.Key)
                        if ($keyInfo.Key -eq 'Y') {
                            $PrBodyContentReply += $_.Replace("[ ]","[X]"), "`n"
                        } else {
                            $PrBodyContentReply += $_, "`n"
                        }
                    }
                }
    
                '*tested your manifest*' {
                    if ($SandboxTest -eq '0') {
                        $PrBodyContentReply += $_.Replace("[ ]","[X]"), "`n"
                    } else {
                        Write-Host
                        Write-Host
                        Write-Host -ForegroundColor 'Yellow' "You did not test your Manifest in Windows Sandbox previously."
                        Write-Host -ForegroundColor 'White' "Have you tested your manifest locally with 'winget install --manifest <path>'"
                        Write-Host -ForegroundColor 'White' -NoNewline "[Y] Yes  "
                        Write-Host -ForegroundColor 'Yellow' -NoNewline "[N] No "
                        Write-Host -NoNewline "(default is 'N'): "
                        do {
                            $keyInfo = [Console]::ReadKey($false)
                        } until ($keyInfo.Key)
                        if ($keyInfo.Key -eq 'Y') {
                            $PrBodyContentReply += $_.Replace("[ ]","[X]"), "`n"
                        } else {
                            $PrBodyContentReply += $_, "`n"
                        }
                    }
                }
    
                '*schema*' {
                    Write-Host
                    Write-Host
                    Write-Host -ForegroundColor 'White' "Does your manifest conform to the 1.0 schema?"
                    Write-Host "Reference Link: https://github.com/microsoft/winget-cli/blob/master/doc/ManifestSpecv1.0.md"
                    Write-Host -ForegroundColor 'White' -NoNewline "[Y] Yes  "
                    Write-Host -ForegroundColor 'Yellow' -NoNewline "[N] No "
                    Write-Host -NoNewline "(default is 'N'): "
                    do {
                        $keyInfo = [Console]::ReadKey($false)
                    } until ($keyInfo.Key)
                    if ($keyInfo.Key -eq 'Y') {
                        $PrBodyContentReply += $_.Replace("[ ]","[X]"), "`n"
                    } else {
                        $PrBodyContentReply += $_, "`n"
                    }
                }
    
                Default {
                    Write-Host
                    Write-Host
                    Write-Host -ForegroundColor 'White' $_.TrimStart("- [ ]")
                    Write-Host -ForegroundColor 'White' -NoNewline "[Y] Yes  "
                    Write-Host -ForegroundColor 'Yellow' -NoNewline "[N] No "
                    Write-Host -NoNewline "(default is 'N'): "
                    do {
                        $keyInfo = [Console]::ReadKey($false)
                    } until ($keyInfo.Key)
                    if ($keyInfo.Key -eq 'Y') {
                        $PrBodyContentReply += $_.Replace("[ ]","[X]"), "`n"
                    } else {
                        $PrBodyContentReply += $_, "`n"
                    }
                }
            }
        }
```

OfficialEsco reviewed (changes requested):
I fixed the Pull request prompt issue, it apparently needed escape characters

I just noticed you have a check if Sandbox was run, but i don't remember seeing that working, the variable is most likely not remembered as it was run in a different Function, you need to change that to $script:SandboxTest for it to be remembered, this is most likely the issue the validate prompt have too

vedantmgoyal2009 commented:
Any method to get PFN of MSIX/APPX file using a command?

OfficialEsco commented:
PFN?

Also about FamilyName and SignatureSha, i've searched around for this before, and from what i found is that it has very little information and support, FamilyName can be found by installing the software and doing get-appxpackage, SignatureSha i have no idea how to read, but for some reason they built it into `winget hash`

vedantmgoyal2009 commented:
Yes, I want to get the package family name without installing, any way to do that?

denelon commented:
@vedantmgoyal2009 as you make improvements/suggestions to YamlCreate, would you mind considering adding new features to the [Windows Package Manager Manifest Creator](https://github.com/microsoft/winget-create)? We will be making this an optional package for the Windows Package Manager so one can `winget create`. We do want to have full fidelity. There may already be sample code for doing some of the things you're asking about.

denelon commented:
The wingetcreate.exe [detects](https://github.com/microsoft/winget-create/blob/4e7f4cdb773c24caf07eccd37d63cfaed501d7c0/src/WingetCreateCore/Common/PackageParser.cs#L626) many of these properties.

vedantmgoyal2009 commented:
> PFN?

Abbreviation for "Package Family Name"

> Also about FamilyName and SignatureSha, i've searched around for this before, and from what i found is that it has very little information and support, FamilyName can be found by installing the software and doing get-appxpackage, SignatureSha i have no idea how to read, but for some reason they built it into `winget hash`

We can at least give a prompt if a user knows how to find it/already have it, then he can fill it while creating the manifests. A while ago, I had to create a manifest in which `SignatureSha265` and `PackageFamilyName` have to be filled but since YamlCreate does not prompt for it, I had to edit and add them manually. If in future, somebody finds a method to automatically get it, we can show it like in ProductCode we do.
```
Product Code found from the installer: <automatically detected product code>
```

> @vedantmgoyal2009 as you make improvements/suggestions to YamlCreate, would you mind considering adding new features to the [Windows Package Manager Manifest Creator](https://github.com/microsoft/winget-create)? We will be making this an optional package for the Windows Package Manager so one can `winget create`. We do want to have full fidelity. There may already be sample code for doing some of the things you're asking about.

I would definitely like to contribute to `wingetcreate` but it will take some time as I will have to understand C#. I currently do not know even 'a' of CSharp.

OfficialEsco commented on Tools/YamlCreate.ps1:
```diff
@@ -786,6 +786,149 @@ Function Test-Manifest {
+                    Write-Host
+                    Write-Host
+                    Write-Host -ForegroundColor 'White' $_.TrimStart("- [ ]")
+                    Write-Host -ForegroundColor 'White' -NoNewline "[Y] Yes  "
+                    Write-Host -ForegroundColor 'Yellow' -NoNewline "[N] No "
+                    Write-Host -NoNewline "(default is 'N'): "
+                    do {
+                        $keyInfo = [Console]::ReadKey($false)
+                    } until ($keyInfo.Key)
+                    if ($keyInfo.Key -eq 'Y') {
+                        $PrBodyContentReply += $_.Replace("[ ]","[X]"), "`n"
+                    } else {
+                        $PrBodyContentReply += $_, "`n"
+                    }
+                }
+            }
+        }
+        Write-Host
+        Write-Host
+        Write-Host -ForegroundColor 'White' "Does this pull request resolve any issues?"
+        Write-Host -ForegroundColor 'White' -NoNewline "[Y] Yes  "
+        Write-Host -ForegroundColor 'Yellow' -NoNewline "[N] No "
+        Write-Host -NoNewline "(default is 'N'): "
+        do {
+            $keyInfo = [Console]::ReadKey($false)
+        } until ($keyInfo.Key)
+        if ($keyInfo.Key -eq 'Y') {
+            Write-Host
+            Write-Host "Enter issue number. For example`: 21983, 43509"
+            $ResolvedIssues = Read-Host -Prompt 'Resolved Issues'
+            Foreach($i in ($ResolvedIssues.Split(",").Trim())) {
+                $PrBodyContentReply += "Resolves #$i`n"
+            }
+        } else {Write-Host}
+        $PrBodyContentReply = ($PrBodyContentReply.Trim() -ne '')
+        Set-Content -Path PrBodyFile -Value $PrBodyContentReply | Out-Null
+        gh pr create --body-file PrBodyFile -f
+        Remove-Item PrBodyFile     
+}
+
```
Found out why it looked weird, accidently made a extra tab space in the Suggested Change

```suggestion
Function Enter-PR-Parameters {
    $PrBodyContent = Get-Content $args[0]
    ForEach ($_ in ($PrBodyContent | Where-Object { $_ -like '-*[ ]*' })) {
        switch -Wildcard ( $_ )
        {
            '*CLA*' {
                Write-Host
                Write-Host -ForegroundColor 'White' "Have you signed the Contributor License Agreement (CLA)?"
                Write-Host "Reference Link: https://cla.opensource.microsoft.com/microsoft/winget-pkgs"
                Write-Host -ForegroundColor 'White' -NoNewline "[Y] Yes  "
                Write-Host -ForegroundColor 'Yellow' -NoNewline "[N] No "
                Write-Host -NoNewline "(default is 'N'): "
                do {
                    keyInfo = [Console]::ReadKey($false)
                } until ($keyInfo.Key)
                if ($keyInfo.Key -eq 'Y') {
                    $PrBodyContentReply += $_.Replace("[ ]","[X]"), "`n"
                } else {
                    $PrBodyContentReply += $_, "`n"
                }
            }
    
            '*open `[pull requests`]*' {
                Write-Host
                Write-Host
                Write-Host -ForegroundColor 'White' "Have you checked that there aren't other open pull requests for the same manifest update/change?"
                Write-Host "Reference Link: https://github.com/microsoft/winget-pkgs/pulls"
                Write-Host -ForegroundColor 'White' -NoNewline "[Y] Yes  "
                Write-Host -ForegroundColor 'Yellow' -NoNewline "[N] No "
                Write-Host -NoNewline "(default is 'N'): "
                do {
                    $keyInfo = [Console]::ReadKey($false)
                } until ($keyInfo.Key)
                if ($keyInfo.Key -eq 'Y') {
                    $PrBodyContentReply += $_.Replace("[ ]","[X]"), "`n"
                } else {
                    $PrBodyContentReply += $_, "`n"
                }
            }
    
            '*winget validate*' {
                if($?) {
                    $PrBodyContentReply += $_.Replace("[ ]","[X]"), "`n"
                } else {
                    Write-Host
                    Write-Host
                    Write-Host -ForegroundColor 'Red' "Automatic manifest validation failed. Check your manifest and try again"
                    Write-Host -ForegroundColor 'White' "Have you validated your manifest locally with 'winget validate --manifest <path>'"
                    Write-Host -ForegroundColor 'White' -NoNewline "[Y] Yes  "
                    Write-Host -ForegroundColor 'Yellow' -NoNewline "[N] No "
                    Write-Host -NoNewline "(default is 'N'): "
                    do {
                        $keyInfo = [Console]::ReadKey($false)
                    } until ($keyInfo.Key)
                    if ($keyInfo.Key -eq 'Y') {
                        $PrBodyContentReply += $_.Replace("[ ]","[X]"), "`n"
                    } else {
                        $PrBodyContentReply += $_, "`n"
                    }
                }
            }
    
            '*tested your manifest*' {
                if ($script:SandboxTest -eq '0') {
                    $PrBodyContentReply += $_.Replace("[ ]","[X]"), "`n"
                } else {
                    Write-Host
                    Write-Host
                    Write-Host -ForegroundColor 'Yellow' "You did not test your Manifest in Windows Sandbox previously."
                    Write-Host -ForegroundColor 'White' "Have you tested your manifest locally with 'winget install --manifest <path>'"
                    Write-Host -ForegroundColor 'White' -NoNewline "[Y] Yes  "
                    Write-Host -ForegroundColor 'Yellow' -NoNewline "[N] No "
                    Write-Host -NoNewline "(default is 'N'): "
                    do {
                        $keyInfo = [Console]::ReadKey($false)
                    } until ($keyInfo.Key)
                    if ($keyInfo.Key -eq 'Y') {
                        $PrBodyContentReply += $_.Replace("[ ]","[X]"), "`n"
                    } else {
                        $PrBodyContentReply += $_, "`n"
                    }
                }
            }
    
            '*schema*' {
                Write-Host
                Write-Host
                Write-Host -ForegroundColor 'White' "Does your manifest conform to the 1.0 schema?"
                Write-Host "Reference Link: https://github.com/microsoft/winget-cli/blob/master/doc/ManifestSpecv1.0.md"
                Write-Host -ForegroundColor 'White' -NoNewline "[Y] Yes  "
                Write-Host -ForegroundColor 'Yellow' -NoNewline "[N] No "
                Write-Host -NoNewline "(default is 'N'): "
                do {
                    $keyInfo = [Console]::ReadKey($false)
                } until ($keyInfo.Key)
                if ($keyInfo.Key -eq 'Y') {
                    $PrBodyContentReply += $_.Replace("[ ]","[X]"), "`n"
                } else {
                    $PrBodyContentReply += $_, "`n"
                }
            }
    
            Default {
                Write-Host
                Write-Host
                Write-Host -ForegroundColor 'White' $_.TrimStart("- [ ]")
                Write-Host -ForegroundColor 'White' -NoNewline "[Y] Yes  "
                Write-Host -ForegroundColor 'Yellow' -NoNewline "[N] No "
                Write-Host -NoNewline "(default is 'N'): "
                do {
                    $keyInfo = [Console]::ReadKey($false)
                } until ($keyInfo.Key)
                if ($keyInfo.Key -eq 'Y') {
                    $PrBodyContentReply += $_.Replace("[ ]","[X]"), "`n"
                } else {
                    $PrBodyContentReply += $_, "`n"
                }
            }
        }
    }

    Write-Host
    Write-Host
    Write-Host -ForegroundColor 'White' "Does this pull request resolve any issues?"
    Write-Host -ForegroundColor 'White' -NoNewline "[Y] Yes  "
    Write-Host -ForegroundColor 'Yellow' -NoNewline "[N] No "
    Write-Host -NoNewline "(default is 'N'): "
    do {
        $keyInfo = [Console]::ReadKey($false)
    } until ($keyInfo.Key)
    if ($keyInfo.Key -eq 'Y') {
        Write-Host
        Write-Host "Enter issue number. For example`: 21983, 43509"
        $ResolvedIssues = Read-Host -Prompt 'Resolved Issues'
        Foreach($i in ($ResolvedIssues.Split(",").Trim())) {
            $PrBodyContentReply += "Resolves #$i`n"
        }
    } else {Write-Host}

    $PrBodyContentReply = ($PrBodyContentReply.Trim() -ne '')
    Set-Content -Path PrBodyFile -Value $PrBodyContentReply | Out-Null
    gh pr create --body-file PrBodyFile -f
    Remove-Item PrBodyFile     
}
```

OfficialEsco reviewed (changes requested):
Also you need to change this
```
switch ($keyInfo.Key) {
            'Y' { $SandboxTest = '0' }
            'N' { $SandboxTest = '1' }
            default { $SandboxTest = '0' }
        }
```
to this
```
switch ($keyInfo.Key) {
            'Y' { $script:SandboxTest = '0' }
            'N' { $script:SandboxTest = '1' }
            default { $script:SandboxTest = '0' }
        }
```
for `if ($script:SandboxTest -eq '0') {` to work since its in another function

OfficialEsco commented:
Do this https://github.com/Trenly/winget-pkgs/pull/4#pullrequestreview-724810079, i can't suggest changes since its outside of the changed scope

OfficialEsco reviewed (approved):
As far as I'm aware this is 👌
